include/json/config.h

```diff
@@ -67,21 +67,40 @@
 // Storages, and 64 bits integer support is disabled.
 // #define JSON_NO_INT64 1
 
-#if defined(_MSC_VER) && _MSC_VER <= 1200 // MSVC 6
-// Microsoft Visual Studio 6 only support conversion from __int64 to double
-// (no conversion from unsigned __int64).
-#define JSON_USE_INT64_DOUBLE_CONVERSION 1
-// Disable warning 4786 for VS6 caused by STL (identifier was truncated to '255'
-// characters in the debug information)
-// All projects I've ever seen with VS6 were using this globally (not bothering
-// with pragma push/pop).
-#pragma warning(disable : 4786)
-#endif // if defined(_MSC_VER)  &&  _MSC_VER < 1200 // MSVC 6
-
-#if defined(_MSC_VER) && _MSC_VER >= 1500 // MSVC 2008
-/// Indicates that the following function is deprecated.
-#define JSONCPP_DEPRECATED(message) __declspec(deprecated(message))
-#endif
+#if defined(_MSC_VER) // MSVC
+	#if _MSC_VER <= 1200 // MSVC 6
+		// Microsoft Visual Studio 6 only support conversion from __int64 to double
+		// (no conversion from unsigned __int64).
+		#define JSON_USE_INT64_DOUBLE_CONVERSION 1
+		// Disable warning 4786 for VS6 caused by STL (identifier was truncated to '255'
+		// characters in the debug information)
+		// All projects I've ever seen with VS6 were using this globally (not bothering
+		// with pragma push/pop).
+		#pragma warning(disable : 4786)
+	#endif // MSVC 6
+
+	#if _MSC_VER >= 1500 // MSVC 2008
+		/// Indicates that the following function is deprecated.
+		#define JSONCPP_DEPRECATED(message) __declspec(deprecated(message))
+	#endif
+
+	#if _MSC_VER >= 1600 // MSVC 2010
+		#define JSON_HAS_RVALUE_REFERENCES
+	#endif 
+#endif // defined(_MSC_VER)
+
+
+#ifdef __clang__
+	#if __has_feature(cxx_rvalue_references)
+		#define JSON_HAS_RVALUE_REFERENCES
+	#endif 
+#elif defined __GNUC__
+	#if defined(__GXX_EXPERIMENTAL_CXX0X__) 
```

Comment:
Perhaps we can `#define JSON_HAS_RVALUE_REFERENCES` to 1 or 0 but only if it's not previously defined. Then we give the end user the option to explicitly `#define` the correct value so that it can be used with all compilers. 
Obviously this would mean using `#if` instead of `#ifdef` when checking conditions.